File: leetcode/726-number-of-atoms/solution.py
"""
Given a chemical formula (given as a string), return the count of each atom. An atomic element
always starts with an uppercase character, then zero or more lowercase letters, representing the name.
1 or more digits representing the count of that element may follow if the count is greater than 1. If
the count is 1, no digits will follow. For example, H2O and H2O2 are possible, but H1O2 is impossible.

Two formulas concatenated together produce another formula. For example, H2O2He3Mg4 is also a formula.

A formula placed in parentheses, and a count (optionally added) is also a formula. For example, (H2O2) and (H2O2)3 are formulas.

Given a formula, output the count of all elements as a string in the following form: the first name (in sorted order),
followed by its count (if that count is more than 1), followed by the second name (in sorted order), followed by its
count (if that count is more than 1), and so on.
---------------------------------------------------
I did this with @ozan during a mock interview. My initial approach was:
- maintain a symbol stack
- get each token (an atomic symbol, a number, or a paren) from the input formula
  - push the symbol onto the stack (casting "12" to literal 12 first) unless it's a closed paren
  - if it's a closed paren, get the number that follows and either append it to every symbol
    on the stack or multiply the existing symbol's count by it; then push everything back on the stack
- finally, read out the stack, obtain counts of symbols, and cast to the final string in sorted order.

Oz had a better idea:
- maintain hashtables like "frames"; start with an empty one and add atomic symbol counts as we parse tokens from the string
- if we encounter a paren, push the current hashtable and start a new one; get the counts til we hit the close paren
- multiply the number following the close paren by everything in our current frame, then merge with the previous frame.
- at the end, read out the keys of our "top level frame" in sorted order

I will implement oz's idea here.
"""
import collections


class Solution:
    def isInt(self, token):
        try:
            return type(int(token)) == int
        except ValueError:
            return False

    def getNextToken(self, formula, idx):
        if idx >= len(formula):
            return ""

        initialIdx = idx
        token = formula[idx]
        if token in ['(', ')']:
            return token
        elif str.isdigit(token):
            idx += 1
            while idx < len(formula) and str.isdigit(formula[idx]):
                token += formula[idx]
                idx += 1
        else:
            if idx < len(formula)-1 and str.islower(formula[idx+1]):
                token += formula[idx+1]
        return token

    def countOfAtoms(self, formula):
        idx = 0
        stack = []
        frame = collections.Counter()
        while idx < len(formula):
            token = self.getNextToken(formula, idx)
            idx += len(token)
            if token == '(':
                stack.append(frame)
                frame = collections.Counter()
            elif token == ')':
                # Get following counter if it exists, multiply
                # counts in existing frame, merge with
                # previous frame
                count = self.getNextToken(formula, idx)
                if count != "":
                    idx += len(count)
                    for key in frame:
                        frame[key] *= int(count)
                stack[-1] += frame
                frame = stack.pop()
            else:
                # We either got an atomic symbol or a number, and
                # numbers only come after symbols
                nextToken = self.getNextToken(formula, idx)
                if self.isInt(nextToken):
                    idx += len(nextToken)
                    frame[token] += int(nextToken)
                else:
                    frame[token] += 1

        solution = ""
        for key in sorted(frame.keys()):
            solution += key
            if frame[key] > 1:
                solution += str(frame[key])
        return solution


if __name__ == '__main__':
    s = Solution()
    assert s.countOfAtoms("") == ""
    assert s.countOfAtoms("H2O") == "H2O"
    assert s.countOfAtoms("Mg(OH)2") == "H2MgO2"
    assert s.countOfAtoms("K4(ON(SO3)2)2") == "K4N2O14S4"
    assert s.countOfAtoms("H50") == "H50"
    assert s.countOfAtoms("(NB3)33") == "B99N33"
    assert s.countOfAtoms("F(NB3)33") == "B99FN33"

File: leetcode/726-number-of-atoms/solution.cpp
/*
Given a chemical formula (given as a string), return the count of each atom.

An atomic element always starts with an uppercase character, then zero or more
lowercase letters, representing the name.

1 or more digits representing the count of that element may follow if the count
is greater than 1. If the count is 1, no digits will follow. For example, H2O
and H2O2 are possible, but H1O2 is impossible.

Two formulas concatenated together produce another formula. For example,
H2O2He3Mg4 is also a formula.

A formula placed in parentheses, and a count (optionally added) is also a
formula. For example, (H2O2) and (H2O2)3 are formulas.

Given a formula, output the count of all elements as a string in the following
form: the first name (in sorted order), followed by its count (if that count is
more than 1), followed by the second name (in sorted order), followed by its
count (if that count is more than 1), and so on.

Example 1:
Input:
formula = "H2O"
Output: "H2O"
Explanation:
The count of elements are {'H': 2, 'O': 1}.

Example 2:
Input:
formula = "Mg(OH)2"
Output: "H2MgO2"
Explanation:
The count of elements are {'H': 2, 'Mg': 1, 'O': 2}.

Example 3:
Input:
formula = "K4(ON(SO3)2)2"
Output: "K4N2O14S4"
Explanation:
The count of elements are {'K': 4, 'N': 2, 'O': 14, 'S': 4}.

Note:
- All atom names consist of lowercase letters, except for the first character
which is uppercase.
- The length of formula will be in the range [1, 1000].
- $formula will only consist of letters, digits, and round parentheses, and is a
valid formula as defined in the problem.
------------------------------------------------
in: string, molecular formula
out: string, same formula with no parens or repeated elements

- parse()
- To keep track of the overall count of each molecule, we need to maintain
a map of symbols to characters. This starts as empty, and at the end of the
function can be turned into a string
- The parsing logic should not modify the string, so it needs to keep
track of its position in the string
- Parens can be handled by recursing, creating a new local context,
parsing, then combining each symbol's count with the multiplier,
and returning / merging with the global context
- We can either keep references to the global context, or return it;
if we return it, we can't return the last index.
- Tokenizing is the tricky part:
  - A capital letter can be followed by:
    - Another capital, meaning the first one is one molecule
    - A lowercase letter, meaning both letters are one molecule
    - A number, indicating count
    - A paren


- After last parse, convert():
  - takes dict, returns string of symbols
*/

#include <iostream>
#include <map>
#include <string>

// Don't do this
using namespace std;

class Solution {
 public:
  /*
   map<string, int> mergeDicts(map<string, int> m1, map<string, int> m2) {}

   string dictToFormula(map<string, int> m) {}

   int parse(string formula, map<string, int> *globalCtx, int i) {
     // map<string, int> localCtx;
     return 0;
   }

   string countOfAtoms(string formula) {
     map<string, int> globalCtx;

     // Don't need returned index;
     parse(formula, &globalCtx, 0);
     return dictToFormula(globalCtx);
   }
   */
};

int main() {
  string str{"ABCD"};
  while (!str.empty()) {
    cout << str[str.size() - 1] << endl;
    str.pop_back();
  }
  return 0;
}
